File: python_implementation/logging/AbstractLogger.cs
using System;
using System.Collections.Generic;
using Codecool.Logging;

namespace Codecool.MainProduct.Logging
{
    /// <summary>
    /// An abstract implementation of the {@code Logger} interface.
    /// Holds some common logic.
    /// Extending classes should override the {@code safeLog} method which
    /// hides the filtering logic for the different log levels.
    /// </summary>
    public abstract class AbstractLogger : ILogger
    {
        private readonly HashSet<LogLevel> _levels;

        /// <summary>
        /// Initializes a new instance of the <see cref="AbstractLogger"/> class.
        /// </summary>
        /// <param name="levels">level parameters</param>
        public AbstractLogger(params LogLevel[] levels)
        {
            if (levels == null)
            {
                throw new ArgumentNullException(nameof(levels));
            }

            _levels = new HashSet<LogLevel>(levels);
        }

        /// <inheritdoc />
        public void Log(LogLevel level, string message)
        {
            WithLevel(level, () => SafeLog(level, message));
        }

        /// <summary>
        /// Log <paramref name="message"/> in a safe manner.
        /// </summary>
        /// <param name="level">Severity level.</param>
        /// <param name="message">Message to log.</param>
        protected abstract void SafeLog(LogLevel level, string message);


        /// <summary>
        /// Invoke <paramref name="logic"/> if the specified <paramref name="level"/> registered.
        /// </summary>
        /// <param name="level">Severity level.</param>
        /// <param name="logic">Actionable method</param>
        private void WithLevel(LogLevel level, Action logic)
        {
            if (_levels.Contains(level))
            {
                logic();
            }
        }
    }
}

File: python_implementation/logging/abstract_logger.py
#!/usr/bin/env python

from enum import Enum
from typing import Callable
from log_level import LogLevel
from ilogger import ILogger


class AbstractLogger(ILogger):
    """
    An abstract implementation of the {@code Logger} interface.
    Holds some common logic.
    Extending classes should override the {@code safeLog} method which
    hides the filtering logic for the different log levels.
    """

    def __init__(self, levels: LogLevel):
        """
        Initializes a new instance of the AbstractLogger class.
        """

        if (levels == None):
            raise ArgumentNullException()

        self._levels = levels

    def log(self, level: LogLevel, message: str) -> None:
        """

        """
        self.__with_level(level, message)


    def safe_log(self, level: LogLevel, message: str) -> None:
        """
        Log in a safe manner.
        """
        raise NotImplementedError()

    def __with_level(self, level: LogLevel, message: str) -> None:
        """

        """
        if level in LogLevel.__members__.values():
            self.safe_log(level, message)
